### PR TITLE
[performance] added network TCP throughput test

### PR DESCRIPTION
**Pull request by iulianbarbu**
## Reason for This PR

Lack of TCP network throughput performance test.

## Description of Changes

Added a test for TCP network throughput for 1/2 vCPUs uVMs. The test contains:
- guest to host workloads
- host to guest workloads
- bidirectional workloads

## License Acceptance

By submitting this pull request, I confirm that my contribution is made under
the terms of the Apache 2.0 license.

## PR Checklist

`[Author TODO: Meet these criteria.]`
`[Reviewer TODO: Verify that these criteria are met. Request changes if not]`

- [x] All commits in this PR are signed (`git commit -s`).
- [x] The reason for this PR is clearly provided (issue no. or explanation).
- [x] The description of changes is clear and encompassing.
- ~~[] Any required documentation changes (code and docs) are included in this PR.~~
- ~~[] Any newly added `unsafe` code is properly documented.~~
- ~~[] Any API changes are reflected in `firecracker/swagger.yaml`.~~
- ~~[] Any user-facing changes are mentioned in `CHANGELOG.md`.~~
- ~~[] All added/changed functionality is tested.~~

### CODE REVIEW

File: tests/integration_tests/performance/test_network_tcp_throughput.py
Comment (raduiliescu) [inline tests/integration_tests/performance/test_network_tcp_throughput.py:35]: I think you have too many options for window size. Also you should leave it to run with default params, including packet length.

File: tests/integration_tests/performance/test_network_tcp_throughput.py
Comment (raduiliescu) [inline tests/integration_tests/performance/test_network_tcp_throughput.py:23]: time is too small, TCP takes some time to get to the best window size. I recommend at least 15s omitting first 4-5s.

File: tests/integration_tests/performance/test_network_tcp_throughput.py
Comment (raduiliescu) [inline tests/integration_tests/performance/test_network_tcp_throughput.py:39]: this is hard to read... I also suggest to put it in a different file so it can be generated and name them goldvalues, or something more suggestive.
Is pretty hard to understand and change them. I recommend you add a mode where you can generate the gold for example a command like ./test_throughput --generate_goldvalues --delta=0.1 ... which would generate the files.

File: tests/integration_tests/performance/test_network_tcp_throughput.py
Comment (raduiliescu) [inline tests/integration_tests/performance/test_network_tcp_throughput.py:157]: While the tests are ok, they are missing a point - how does the emulation perform in a high throughput scenarios.
Do these tests fill up the emulation? How does the performance look when you generate a lot of contention?
I recommend to run tests with different vm sizes - 1vcpus, 2, 4, 8, 16 and also try multiple iperfs processes for each vm type 1, 2, 4, 8, 16, 128. 
Also after this PR please try with multiple rate-limiter configs, so we can know the penalty added by the rate-limiter.

Comment (iulianbarbu) [reply tests/integration_tests/performance/test_network_tcp_throughput.py:235]: Trying with different uVMs machine configs definitely makes sense. Adding these scenarios is straight forward, starting from this test, but I suggest we keep this as it is right now, following up with uVMs of 4, 8, 16, 128 vCPUs in the future. As we have seen for what we have right now, getting baselines for these uVMs take some time. Once this PR is merged, it is just a matter of adding a new microvm artifact and getting the baselines. WDYT?

File: tests/integration_tests/performance/test_network_tcp_throughput.py
Comment (iulianbarbu) [reply tests/integration_tests/performance/test_network_tcp_throughput.py]: Reduced the options for window size. PTAL!

File: tests/integration_tests/performance/test_network_tcp_throughput.py
Comment (iulianbarbu) [reply tests/integration_tests/performance/test_network_tcp_throughput.py]: Raised the workload runtime to 20 seconds.

File: tests/integration_tests/performance/test_network_tcp_throughput.py
Comment (iulianbarbu) [reply tests/integration_tests/performance/test_network_tcp_throughput.py]: From my experience of running this test, generating gold values is a process which should run the test for many iterations, extracting the minimum and the maximum for each test combination. The gold values contains a target and a delta. The target can be the mean: (MIN + MAX) / 2 and delta: (MAX - MIN) / 2 + a not significant buffer space.

This means that we can not simply overwrite the baselines, running the test once. We do not have an assurance that this process will provide the best resilience to intermittent failures.

I've improved the information presented in the baseline. Please let me know if you still find the baselines hard to read.

**Comment by dianpopa** [comment]
@iulianbarbu is this a duplicate of https://github.com/firecracker-microvm/firecracker/pull/1877? Do we think we can close it?

**Comment by iulianbarbu** [comment]
> Not clear do you also run with default params for window size and packet size? If not please make it clear here with adding something like 0 in this list, otherwise I don't see how you can handle it generically

I test close to default params, but not explicitly the defaults. The default params depend on the configured socket buffer size, on the OS distribution we run. I'll make it more obvious that we test the defaults.

**Comment by raduiliescu** [comment]
> 
> 
> > Not clear do you also run with default params for window size and packet size? If not please make it clear here with adding something like 0 in this list, otherwise I don't see how you can handle it generically
> 
> I test close to default params, but not explicitly the defaults. The default params depend on the configured socket buffer size, on the OS distribution we run. I'll make it more obvious that we test the defaults.

That is not ok. You should run without a strict window size so it can autoadjust according to the TCP algorithms.
Also for the packet size default it will run with maximum one.

File: tests/integration_tests/performance/network_tcp_throughput_test_config.py
Comment (raduiliescu) [inline tests/integration_tests/performance/network_tcp_throughput_test_config.py:33]: I don't get what you did now. Is really important to test with small packets... My suggestion was to keep old values and add a None...

Comment (iulianbarbu) [reply tests/integration_tests/performance/network_tcp_throughput_test_config.py:33]: The previous packet length was 128K, which is the default packet length used by iperf3 for TCP packets. So there is no change in terms of packet size. The value for 256K related to window size, which I've replaced with `None`, is very similar to the results obtained by using the default window size (or more specifically, as TCP decides). Please let me know if I misunderstood something.

File: tests/host_tools/network.py
Comment (serban300) [inline tests/host_tools/network.py:41]: What's the difference between executing a command with asyncio and without asyncio ? Why do we need this ?

File: tests/framework/statistics/criteria.py
Comment (serban300) [inline tests/framework/statistics/criteria.py:10]: Another option would be to throw an Exception

File: tests/framework/utils.py
Comment (serban300) [inline tests/framework/utils.py:400]: what does `cp` stand for ?

File: tests/framework/utils.py
Comment (serban300) [inline tests/framework/utils.py:423]: I would test if `cpu_load[info[1]]` is not initialized. Otherwise there might be corner cases. For example if we get here before Firecracker fully started, the vcpu threads won't exist, so we won't initialized the list. And after a couple of iterations the vcpu threads will have started.

File: tests/integration_tests/performance/test_network_tcp_throughput.py
Comment (serban300) [reply tests/integration_tests/performance/test_network_tcp_throughput.py]: I would rename it to something like `iperf_index` or `host_iperf_index`.

File: tests/integration_tests/performance/network_tcp_throughput_test_config.py
Comment (raduiliescu) [inline tests/integration_tests/performance/network_tcp_throughput_test_config.py:32]: Please add also a small packet size 128 or 96. The 128K is not the packet size, but the payload size that is later segmented, since nobody is sending a 128K packet on the wire being limited first by mss at tcp level and finaly but MTU. The window size looks good now

File: tests/integration_tests/performance/test_network_tcp_throughput.py
Comment (raduiliescu) [reply tests/integration_tests/performance/test_network_tcp_throughput.py:26]: Please track this remaining requirement

File: tests/host_tools/network.py
Comment (iulianbarbu) [reply tests/host_tools/network.py:41]: The issue is not that the command ran on the same threads. In practice, `run_cmd` will start iperf3 server processes in the background (on the host), on the main thread, and we do not account for them. We account for the guest processes, started with `execute_command`. Those commands have to be tracked in order to get their `stdout`. This has to happen on different threads, while each thread monitors a specific iperf3 client stdout and returns it to the main thread. Then the issue here is that the inner implementation of `execute_command` depends on asyncio runtime, which provides an incompatible interface to be used within threads, because of [1].

[1] https://python.readthedocs.io/en/latest/library/asyncio-dev.html#concurrency-and-multithreading

File: tests/integration_tests/performance/test_network_tcp_throughput.py
Comment (iulianbarbu) [reply tests/integration_tests/performance/test_network_tcp_throughput.py:81]: For a `load_factor` of one this will probably assign the same CPU core to a FC thread and to iperf3 server processes as well. I try to avoid this by having this method aware of other processes/threads already pinned to CPU cores, which are not available anymore.

Comment (serban300) [reply tests/integration_tests/performance/test_network_tcp_throughput.py:81]: Right, sorry.

**Review by iulianbarbu — COMMENTED** [review]
I just need to update baselines for the new payload length of 96K. I think I've addressed the rest of your comments. @raduiliescu , @serban300

File: tests/framework/utils.py
Comment (serban300) [reply tests/framework/utils.py:427]: Hmmm I don't see any change to this `if`. It's still 

```
if i == 0:
    cpu_load[info[1]] = list()
```

File: tests/host_tools/network.py
Comment (serban300) [reply tests/host_tools/network.py]: We should understand what causes this behaviour. Why running guest commands with asyncio can lead to deadlocks. I don't think this is the desired behavior.

Your code works around the problem for this specific scenario. But we should make sure we don't encounter this issue in other scenarios as well.

**Comment by raduiliescu** [comment]
Now it looks ok to me.

File: tests/framework/utils.py
Comment (iulianbarbu) [reply tests/framework/utils.py]: Sorry, this comment somehow slipped. Please take a look again. The `get_cpu_percent` primitive is changed a bit.

**Comment by iulianbarbu** [comment]
@serban300 , @raduiliescu , I've integrated your feedback, but found difficulty in updating the baselines - hence the multiple force-pushes done so far. At the moment, I am pleased with the pipeline results, based on the current baselines. We can have a final say after the weekend, by looking at the reliability of the pipeline.

PTAL and let me know if you find things I can improve/fix.

**Review by raduiliescu — APPROVED** [review]

**Review by serban300 — APPROVED** [review]

**Review by raduiliescu — APPROVED** [review]

**Pull request merged**
